libbpf-tools/doublefree.bpf.c

```diff
@@ -0,0 +1,158 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2022 LG Electronics */
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include "doublefree.h"
+
+const volatile bool kernel_threads_only = false;
+const volatile bool user_threads_only = false;
+const volatile pid_t targ_tgid = -1;
+const volatile pid_t targ_pid = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, struct doublefree_info_t);
+} allocs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} memptrs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(max_entries, MAX_ENTRIES);
+	__uint(key_size, sizeof(u32));
+} stack_traces SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u32);
+} first_deallocs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u32);
+} second_deallocs SEC(".maps");
+
+int gen_alloc_exit2(struct pt_regs *ctx, u64 address)
+{
+	struct doublefree_info_t info = {0, };
+
+	if (address != 0) {
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces, BPF_F_USER_STACK);
+		info.is_available = 1;
+		bpf_map_update_elem(&allocs, &address, &info, BPF_ANY);
+	}
+	return 0;
+}
+
+int gen_alloc_exit(struct pt_regs *ctx)
+{
+	return gen_alloc_exit2(ctx, PT_REGS_RC(ctx));
+}
+
+int gen_free_enter(struct pt_regs *ctx, void *address)
+{
+	int stack_id = 0;
+	u64 addr = (u64)address;
+	struct doublefree_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+
+	if (info != NULL) {
+		info->is_available -= 1;
+		stack_id = bpf_get_stackid(ctx, &stack_traces, BPF_F_USER_STACK);
+		if (info->is_available == 0) {
+			bpf_map_update_elem(&first_deallocs, &addr, &stack_id, BPF_ANY);
+		} else if (info->is_available < 0) {
+			bpf_map_update_elem(&second_deallocs, &addr, &stack_id, BPF_ANY);
+		} else {
+			bpf_printk("This code should not be printed at any rate\n");
+			bpf_printk("Please check if something goes wrong\n");
+		}
+	}
+	return 0;
+}
+
+SEC("kretprobe/dummy_malloc")
+int BPF_KPROBE(malloc_return)
```

Comment:
BPF_KRETPROBE

Reply:
Thanks, I will change it from BPF_KPROBE into BPF_KRETPROBE.